### PR TITLE
send initial unready endpoints when SendUnhealthyEndpoints is enabled

### PR DESCRIPTION
**Pull request by ramaraochavali**
Fixes https://github.com/istio/istio/issues/46892

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@howardjohn WDYT? I know I have initially [suggested](https://github.com/istio/istio/issues/18367#issuecomment-546789645) not to send "UnReady" endpoints to avoid issues with panic threshold. But now since, we disable panic threshold when `SendUnhealthyEndpoints` is enabled, I think it is safe to send the initial non ready end points. It is also needed to fix the inconsistency between Istiods as described in https://github.com/istio/istio/issues/46892.

Let me know if you have any alternate suggestions.

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/model/endpointshards.go [hunk 1]
```diff
@@ -288,44 +288,48 @@ func (e *EndpointIndex) UpdateServiceEndpoints(
...
+				nmap[nie.Address] = nie
+			}
+			for _, nie := range istioEndpoints {
+				if oie, exists := emap[nie.Address]; exists {
+					// If endpoint exists already, we should push if it's health status changes.
+					if oie.HealthStatus != nie.HealthStatus {
+						needPush = true
+					}
+					newIstioEndpoints = append(newIstioEndpoints, nie)
+				} else {
```
Comment (Purushotham233) [inline pilot/pkg/model/endpointshards.go:319]: Can we use the old state when available? Or an option to configure that would be desireable.

I understand choosing the option may introduce inconsistent behaviour from different istiod pods.

Comment (ramaraochavali) [reply pilot/pkg/model/endpointshards.go:319]: I think it is not desirable to have inconsistent behaviour. The reason why we disabled panic threshold is it leads to the case where a non ready pod gets request which is against k8s.

File: pilot/pkg/model/endpointshards.go [hunk 2]
```diff
@@ -288,44 +288,48 @@ func (e *EndpointIndex) UpdateServiceEndpoints(
...
-		}
-		for _, nie := range istioEndpoints {
-			nmap[nie.Address] = nie
-		}
 		needPush := false
-		for _, nie := range istioEndpoints {
-			if oie, exists := emap[nie.Address]; exists {
-				// If endpoint exists already, we should push if it's health status changes.
-				if oie.HealthStatus != nie.HealthStatus {
+		if oldIstioEndpoints == nil {
```
Comment (Purushotham233) [inline pilot/pkg/model/endpointshards.go:292]: When oldIstioEndpoints are null, given istiod is restarted, how is the execution not entering the below code branch.
```
	ep, created := e.GetOrCreateEndpointShard(hostname, namespace)
	// If we create a new endpoint shard, that means we have not seen the service earlier. We should do a full push.
	if created {
		log.Infof("Full push, new service %s/%s", namespace, hostname)
		pushType = FullPush
	}
```

Comment (Purushotham233) [reply pilot/pkg/model/endpointshards.go:292]: I see, without the change, PushType is geting updatd from FullPush to NoPush.

Would skipping the iteration through the endpoints when PushType is FullPush. 
But that would also make the codeflow different for the fresh istiod pod and the old istiod pod.

**Pull request merged**